File: solutions/go/0002.add-two-numbers/solution.go
// Created by seognil LC at 2023/07/09 22:53
// leetgo: 1.3.3
// https://leetcode.cn/problems/add-two-numbers/

package main

import (
	"bufio"
	"fmt"
	"os"

	. "github.com/j178/leetgo/testutils/go"
)

// * [ '8 ms', '81%', '4.3 MB', '89%' ]

// @lc code=begin

func addTwoNumbers(l1 *ListNode, l2 *ListNode) (ans *ListNode) {

	carry := 0
	dummyHead := &ListNode{}
	currentNode := dummyHead

	for l1 != nil || l2 != nil || carry != 0 {
		val1, val2 := 0, 0
		if l1 != nil {
			val1, l1 = l1.Val, l1.Next
		}
		if l2 != nil {
			val2, l2 = l2.Val, l2.Next
		}

		sum := val1 + val2 + carry
		carry, sum = sum/10, sum%10

		currentNode.Next = &ListNode{Val: sum}
		currentNode = currentNode.Next
	}

	return dummyHead.Next
}

// @lc code=end

func main() {
	stdin := bufio.NewReader(os.Stdin)
	l1 := Deserialize[*ListNode](ReadLine(stdin))
	l2 := Deserialize[*ListNode](ReadLine(stdin))
	ans := addTwoNumbers(l1, l2)

	fmt.Println("\noutput:", Serialize(ans))
}

File: solutions/python3/0002.add-two-numbers/solution.py
# Created by seognil LC at 2023/07/09 22:53
# leetgo: 1.3.3
# https://leetcode.cn/problems/add-two-numbers/

from typing import *
from leetgo_py import *


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# * [ '56 ms', '86%', '15.9 MB', '68%' ]

# @lc code=begin


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        carry = 0
        dummyHead = currentNode = ListNode()

        while l1 or l2 or carry:
            carry += (l1.val if l1 else 0) + (l2.val if l2 else 0)
            currentNode.next = ListNode(carry % 10)
            carry //= 10

            # * a bit slower
            # carry, sum = divmod(
            #     (l1.val if l1 else 0) + (l2.val if l2 else 0) + carry, 10
            # )
            # currentNode.next = ListNode(sum)

            currentNode = currentNode.next

            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        return dummyHead.next


# @lc code=end

if __name__ == "__main__":
    l1: ListNode = deserialize("ListNode", read_line())
    l2: ListNode = deserialize("ListNode", read_line())
    ans = Solution().addTwoNumbers(l1, l2)

    print("\noutput:", serialize(ans))
